### ISSUE TITLE
root TAP should output serialized data for `node --test` runner

### ISSUE THREAD

**Issue by isaacs** [tapjs/tapjs#897]
See #888 

Output serialization events instead of TAP when env.NODE_TEST_CONTEXT=child-v8.

---

Output serialization events instead of `TAP` when
`env.NODE_TEST_CONTEXT=child-v8`.

All the objects are serialized with `{ type, data }`.

`data.file` is always the `t.options.at.fileName` if found, or
`undefined` if the repl or something else.

- `test:dequeue` emit for tap's `subtestStart` event. Data is
  `{name: string, file: string, nesting: number}`. `file`,
  `line`, and `column` come from `t.options.at`, nesting is the
  _parent_ nesting level, with children of the root TAP object
  being nesting level `0`. (No event emitted for root TAP object,
  because it has no parent to emit `subtestStart`.)
- `test:diagnostic` emit for TAP comments. `{file:string,
  message: string, nesting: number }`. `file`, `line`, and
  `column` come from `t.options.at`. It looks like `t.diagnostic`
  comments are printed in the parent after the test ends, so
  again `nestingLevel` applies to the parent. `t.diagnostic()` is
  a bit like `t.after(() => t.comment(msg))`, printing on the
  parent after the test ends.
- `test:enqueue` emit for tap's `subtestAdd` event.
  `{name:string,file:string,nesting:number}` same semantics as
  `test:dequeue`.
- `test:plan` not the actual plan, since `node:test` only expects
  test points for _subtests_, not _assertions_. This is always a
  trailing value, there is no `t.plan()` in `node:test`. Options:
    - emit this serialized event for suites, but not assertions
    - emit `test:enqueue`/`test:dequeue`/`test:pass`/`test:fail`
      for all _assertions_ as well. This is likely to be very
      noisy, especially if the `node --test` harness is
      outputting TAP itself.
- `test:start` emit immediately after `test:dequeue`, with the
  same data.
- `test:stdout` any writes to stdout. `{file, message:string}`
- `test:stdout` any writes to stderr. `{file, message:string}`
- `test:pass` Emitted when a subtest emits `complete` with
  a FinalResults (not just any passing assertion).
- `test:fail` Same as `test:pass`, but if it's failing.

Pass/fail data:

- `details`
    - `duration_ms` what node-tap calls `# time=...`, but always
      in ms.
    - `type` set to `'suite'` if created with `describe()`.
      Literally no other difference that I can tell. Probably
      tap should just set this if the test had any subtests?
    - `error` for `test:fail`, this is an Error object describing
      the failure. See below.
- `file`, `line`, and `column` come from `t.options.at`.
- `name` the name of the test that's finished
- `nesting` the nesting level of the _parent_ test. (not emitted
  for the root TAP test.)
- 

### Failure `details.error`

Node only fails tests by throwing an error. Tap can have 1 or
more failed test points, _or_ a thrown error.

If we throw an error, that's easy, there's 1 failed test point,
and it's error shaped.

If we get multiples what to do? Just serialize the first one as
the error cause?

Error message is always `'[Error [ERR_TEST_FAILURE]: ${failed
test point message}]'`, with a `code: 'ERR_TEST_FAILURE'` and a
`failureType` of `'subtestsFailed'` if caused by a subtest, or
`'testCodeFailure'` if it failed locally.

The `cause` property is set to whatever triggered the failure. In
the case of a thrown error, it's the error that was thrown, but
it can also be any value.

Since failed test points typically have a `stack`, we can
probably create an Error object for each one, and then have
something like, `Object.assign(new Error('7 test points failed'),
{ cause: this.lists.fail })` or something, presenting an array of
data objects pointing to the sources. If there's only one test
point, create an `Error` for it, with the appropriate message and
stack, and serialize that as the `cause` property.

Any other diagnostics that aren't captured in the error (ie,
anything but `cause`, `message`, and `stack`) can be attached as
additional properties, these are displayed.

**Comment by isaacs** [comment]
Slight wrinkle. `node:test` tests are _always_ async, and they are enqueued ahead of time and run in queued order. But, the queuing of subtests only happens once the parent test starts, and is completely finished when the parent test is processed.

So for a simple test like:

```js
import t from 'tap'

t.jobs = 1

t.test('a', async t => {
  t.test('b', async () => {})
  t.test('c', async () => {})
})

t.test('x', async t => {
  t.test('y', async () => {})
  t.test('z', async () => {})
})
```

We see these events:

```
t.ts > TAP > a subtestAdd
t.ts > TAP > a subtestStart
t.ts > TAP > a > b subtestAdd
t.ts > TAP > a > b subtestStart
t.ts > TAP > a > c subtestAdd
t.ts > TAP > x subtestAdd
t.ts > TAP > a > b subtestProcess
t.ts > TAP > a > b complete
t.ts > TAP > a > c subtestStart
t.ts > TAP > a > c subtestProcess
t.ts > TAP > a > c complete
t.ts > TAP > a subtestProcess
t.ts > TAP > a complete
t.ts > TAP > x subtestStart
t.ts > TAP > x > y subtestAdd
t.ts > TAP > x > y subtestStart
t.ts > TAP > x > z subtestAdd
t.ts > TAP > x > y subtestProcess
t.ts > TAP > x > y complete
t.ts > TAP > x > z subtestStart
t.ts > TAP > x > z subtestProcess
t.ts > TAP > x > z complete
t.ts > TAP > x subtestProcess
t.ts > TAP > x complete
```

Which is _almost_ translatable to `node:test` message ordering, but it gets completely screwball if `t.jobs` is high and they're actually delayed:

```js
import t from 'tap'
import { setTimeout } from 'node:timers/promises'

t.jobs = 99

t.test('a', async t => {
  t.jobs = 99
  t.test('b', async () => setTimeout(1))
  t.test('c', async () => setTimeout(1))
})

t.test('x', async t => {
  t.jobs = 99
  t.test('y', async () => setTimeout(1))
  t.test('z', async () => setTimeout(1))
})
```

```
t.ts > TAP > a subtestAdd
t.ts > TAP > a subtestStart
t.ts > TAP > a > b subtestAdd
t.ts > TAP > a > b subtestStart
t.ts > TAP > a > c subtestAdd
t.ts > TAP > a > c subtestStart
t.ts > TAP > x subtestAdd
t.ts > TAP > x subtestStart
t.ts > TAP > x > y subtestAdd
t.ts > TAP > x > y subtestStart
t.ts > TAP > x > z subtestAdd
t.ts > TAP > x > z subtestStart
t.ts > TAP > a > b subtestProcess
t.ts > TAP > a > b complete
t.ts > TAP > a > c subtestProcess
t.ts > TAP > a > c complete
t.ts > TAP > a subtestProcess
t.ts > TAP > a complete
t.ts > TAP > x > y subtestProcess
t.ts > TAP > x > y complete
t.ts > TAP > x > z subtestProcess
t.ts > TAP > x > z complete
t.ts > TAP > x subtestProcess
t.ts > TAP > x complete
```

I think the thing to do here is just build up a tree of subtests, with the appropriate message data captured for each, and emit them all at once in order at the end.

**Comment by isaacs** [comment]
Yeah, that's a lot easier, completely matches node:test message ordering

```
t.ts > TAP > a test:enqueue
t.ts > TAP > x test:enqueue
t.ts > TAP > a test:dequeue
t.ts > TAP > a test:start
t.ts > TAP > a > b test:enqueue
t.ts > TAP > a > c test:enqueue
t.ts > TAP > a > b test:dequeue
t.ts > TAP > a > b test:start
t.ts > TAP > a > b test:pass/fail b { pass: true, suite: false }
t.ts > TAP > a > c test:dequeue
t.ts > TAP > a > c test:start
t.ts > TAP > a > c test:pass/fail c { pass: true, suite: false }
t.ts > TAP > a test:plan 2
t.ts > TAP > a test:pass/fail a { pass: true, suite: true }
t.ts > TAP > x test:dequeue
t.ts > TAP > x test:start
t.ts > TAP > x > y test:enqueue
t.ts > TAP > x > z test:enqueue
t.ts > TAP > x > y test:dequeue
t.ts > TAP > x > y test:start
t.ts > TAP > x > y test:pass/fail y { pass: true, suite: false }
t.ts > TAP > x > z test:dequeue
t.ts > TAP > x > z test:start
t.ts > TAP > x > z test:pass/fail z { pass: true, suite: false }
t.ts > TAP > x test:plan 2
t.ts > TAP > x test:pass/fail x { pass: true, suite: true }
t.ts > TAP test:plan 2
t.ts > TAP test:pass/fail TAP { pass: true, suite: true }
```

**Comment by isaacs** [comment]
getting closer
<img width="528" alt="Screenshot 2023-09-19 at 17 04 26" src="https://github.com/tapjs/tapjs/assets/9287/90bbdfe0-84e5-4b63-9e9b-8b0467f79937">

**Comment by isaacs** [comment]
<img width="824" alt="Screenshot 2023-09-19 at 22 44 01" src="https://github.com/tapjs/tapjs/assets/9287/f338f841-b637-4f08-913d-aa7072ef7fc3">

Ok, time to clean up this spike and get it done, I think. That's about as nice as it's gonna get, probably. The more advanced colored reporting and syntax highlighting stuff is all chalk dependent, so it's esm only.

It feels a bit gross to shove everything into the error message like that, but it seems to work pretty reliably, and whatever, it's a string.

**Comment by isaacs** [comment]
Landed on #888
